WIP: Automatically derive nested pydantic class

thejaminator opened the pull request:
Could we automatically derive nested pydantic classes?
Suppose someone has a repo using pydantic and wants to start using strawberry.
It is troublesome to manually create a new class for each pydantic basemodel and decorate it with `@strawberry.experimental.pydantic.type`. (Thats the issue i am facing now)

Could we automatically (optionally) derive strawberry types for nested pydantic classes?
My initial MR works for this basic example. I'm opening one early so i can find out what issues may occur.


```
from typing import List
import strawberry
from pydantic import BaseModel


class Child(BaseModel):
    value: str


class Parent(BaseModel):
    children: List[Child]


class GrandParent(BaseModel):
    children: List[Parent]


@strawberry.experimental.pydantic.type(model=GrandParent, all_fields=True)
class GrandParentStrawberry: # only need to decorate GrandParentStrawberry
    ...


@strawberry.type
class Query:
    @strawberry.field
    def hello(self) -> GrandParentStrawberry:
        return GrandParentStrawberry.from_pydantic(
            GrandParent(children=[Parent(children=[Child(value="123")])])
        )

```


<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

<!--- Describe your changes in detail here. -->

## Types of Changes

<!--- What types of changes does your pull request introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

*

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

botberry commented:
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to James Chua for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```
<!-- action-check: release-file -->

patrick91 commented:
I'm not opposed to this :) I definitely like the idea of making it optional

Let's see what other people think! @la4de (this might be something we could, if we don't already, do for django), @jkimbo, @BryceBeagle, @marcoacierno

thejaminator commented on strawberry/experimental/pydantic/object_type.py:
```diff
@@ -25,15 +26,30 @@
+
+    format_type: Callable[[str], str] = lambda x: x # Default does not rename the nested model's name
+    # todo: Other options such as exclude all fields that have a certain name
+
```
i didn't put these arguments in the decorator arguments because these options only apply if register_nested is passed. 
format_type is quite useful in my use case because we frequently have to rename "Type" into "Type{ServiceName}".

Alternatively, we could support a bunch of class decorator for pydantic basemodels. which is what the library [Caliban does for scala](https://ghostdogpr.github.io/caliban/docs/schema.html#annotations).

```
@GQLName("UserGQL") # rename its to UserGQL when it gets automatically registered
class User(pydantic.BaseModel):
     ...
```

thejaminator commented:
thanks, 

> I'm not opposed to this :) I definitely like the idea of making it optional
> 
> Let's see what other people think! @la4de (this might be something we could, if we don't already, do for django), @jkimbo, @BryceBeagle, @marcoacierno

thanks, i've fixed the tests and added options for registration of nested basemodels. Let me know if that is desirable and i can update the docs

BryceBeagle commented on strawberry/experimental/pydantic/object_type.py:
```diff
@@ -48,19 +67,38 @@ def replace_pydantic_types(type_: Any):
+
     if issubclass(type_, BaseModel):
         if hasattr(type_, "_strawberry_type"):
             return type_._strawberry_type
         else:
-            raise UnregisteredTypeException(type_)
+            if register_nested is not None:
+                # register nested class
+                type_name = register_nested.format_type(type_.__name__)
+                strawberry.experimental.pydantic.type(
+                    model=type_, all_fields=True, register_nested=register_nested
+                )(create_type(type_name, (), {}))
+
+                assert hasattr(
+                    type_, "_strawberry_type"
+                ), f"Failed to create _strawberry_type for {type_.__name__}. Please report this bug"
+
+                return type_._strawberry_type
+            else:
+                raise UnregisteredTypeException(type_)
```
Logic here can be simplified and # of returns reduced if we reverse the `if/else`, plus I think it reads a bit better:

```python
if not hasattr(type_, "_strawberry_type"):
    ...
return type_._strawberry_type
```

BryceBeagle commented on tests/experimental/pydantic/test_basic.py:
```diff
@@ -187,6 +189,94 @@ class UserType:
+    assert field1.type is int
+
+    assert field2.python_name == "group"
+    assert field2.type.__name__ == "Group", "should not rename type by default"
```
For consistency with the rest of the tests in the repo, can we remove the custom assert messages? If you thin the message is important, let's just make it a comment/docstring on the test.

thejaminator replied:
sure

XChikuX commented:
I know we are working on first class pydantic support. However, this PR would benefit me highly!
Is there any reasonn this hasn't been commited? Other than a release.md ?

patrick91 commented:
> I know we are working on first class pydantic support. However, this PR would benefit me highly! Is there any reasonn this hasn't been commited? Other than a release.md ?

From my side, I don't use Pydantic much so I haven't been able to test it properly :)

@XChikuX would you be able to test this if I make a pre-release? :)

patrick91 commented:
@XChikuX rebasing this might take me a bit of time :D

thejaminator commented:
I'm not really sure if this pattern works appropriately in all cases. Like
what happens if you want to manually add fields in the converted type. How
do you make sure the user has control over it.

On Wed, Jan 18, 2023, 6:22 PM Patrick Arminio ***@***.***>
wrote:

> @XChikuX <https://github.com/XChikuX> rebasing this might take me a bit
> of time :D
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/strawberry-graphql/strawberry/pull/1477#issuecomment-1386816837>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHI2757ZM4QBZA5PMCZIHGTWS675BANCNFSM5JQ3GYLA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

patrick91 commented:
> I'm not really sure if this pattern works appropriately in all cases. Like what happens if you want to manually add fields in the converted type. How do you make sure the user has control over it.

that's a fair point. We need to think more about this pattern. I also see a future we enable pydantic via something like this:

```
@strawberry.type # or pydantic.type
class MyType(BaseModel):
    ...
```

and then in that case it would work since we are exposing all the fields (we'd basically treat pydantic as a first class citizen).

But for how strawberry.experimental.pydantic works, we probably might want to have a different behaviour. Maybe we can add a flag on type (ie convert_nested=True)? it's a bit ugly though

XChikuX commented:
I agree we shouldn't throw this into main without covering the edge cases.
Maybe can just leave it as `nested=True` for now, Conversion should be implied in the one word.

I don't mind waiting for pydantic support, if its coming in the next month or two.
Otherwise, I'd like to push for this since its already half implemented. I can try to cover the edge cases like what @thejaminator mentioned when I get time.

My code is quickly getting verbose using the experimental library.

thejaminator commented:
Sure, go ahead. I'll suggest that instead of a parameter is_nested: bool,
make it something like nested_options: Optional[NestedOptions].
NestedOptions can be a data class that describes the settings for
conversions. That allows us to add more settings in the future.
The None case will indicate our current behavior.


There's a few options that i foresee being useful. Like a callable that
allows you to specify the type renaming behavior. And in the future maybe
something like a set of  Type[BaseModel] that the user can add to exclude
from automatic conversion.

On Wed, Jan 18, 2023, 7:13 PM Srikanth ***@***.***> wrote:

> I agree we shouldn't throw this into main without covering the edge cases.
> Maybe can just leave it as nested=True for now, Conversion should be
> implied in the one word.
>
> I don't mind waiting for pydantic support, if its coming in the next month
> or two.
> Otherwise, I'd like to push for this since its already half implemented. I
> can try to cover the edge cases like what @thejaminator
> <https://github.com/thejaminator> mentioned when I get time.
>
> My code is quickly getting verbose using the experimental library.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/strawberry-graphql/strawberry/pull/1477#issuecomment-1386884230>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHI275Z4LJ367VUDBVSX6MTWS7F5DANCNFSM5JQ3GYLA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

thejaminator commented:
@XChikuX i'll commit to coming up with something during our strawberry sprint on 18 feb :)
